src/text.c

```diff
@@ -0,0 +1,441 @@
+		SdbListIter *it;
+		void *token_off_tmp;
+		ctx->cur_db = ctx->root_db;
+		ls_foreach (ctx->path, it, token_off_tmp) {
+			size_t token_off = (size_t)token_off_tmp;
+			if (!ctx->buf[token_off]) {
+				continue;
+			}
+			ctx->cur_db = sdb_ns (ctx->cur_db, ctx->buf + token_off, 1);
+			if (!ctx->cur_db) {
+				ctx->cur_db = ctx->root_db;
+				break;
+			}
+		}
+		ls_destroy (ctx->path);
+		break;
+	}
+	case STATE_VALUE:
+		if (!*ctx->buf || !ctx->buf[ctx->token_begin]) {
+			break;
+		}
+		sdb_set (ctx->cur_db, ctx->buf, ctx->buf + ctx->token_begin, 0);
+		break;
+	default:
+		break;
+	}
+	if (ctx->buf_filled && ctx->pos < ctx->buf_filled - 1) {
+		memmove (ctx->buf, ctx->buf + ctx->pos + 1, ctx->buf_filled - ctx->pos - 1);
+	}
+	// prepare for next line
+	ctx->shift = 0;
+	if (ctx->pos + 1 <= ctx->buf_filled) {
+		ctx->buf_filled -= ctx->pos + 1;
+	} else {
+		// special case, when at the end of the file
+		// here pos == buf_filled
+		ctx->buf_filled = 0;
+	}
+	ctx->pos = 0;
+	ctx->state = STATE_NEWLINE;
+}
+
+static void load_process_single_char(LoadCtx *ctx) {
+	if (ctx->buf[ctx->pos] == '\n' || ctx->buf[ctx->pos] == '\r') {
+		load_flush_line (ctx);
+		return;
+	}
+
+	if (ctx->state == STATE_NEWLINE) {
+		// at the start of a line, decide whether it's a path or a k=v
+		// by whether there is a leading slash.
+		if (ctx->buf[ctx->pos] == '/') {
+			ctx->state = STATE_PATH;
+			ctx->token_begin = 1;
+			ctx->pos++;
+			return;
+		} else {
+			ctx->state = STATE_KEY;
+		}
+	}
+
+	if (ctx->unescape) {
+		char raw_char;
+		switch (ctx->buf[ctx->pos]) {
+		case 'n':
+			raw_char = '\n';
+			break;
+		case 'r':
+			raw_char = '\r';
+			break;
+		default:
+			raw_char = ctx->buf[ctx->pos];
+		break;
+		}
+		ctx->buf[ctx->pos - ctx->shift] = raw_char;
+		ctx->unescape = false;
+	} else if (ctx->buf[ctx->pos] == '\\') {
+		// got a backslash, the next char, unescape in the next iteration or die!
+		ctx->shift++;
+		ctx->unescape = true;
+	} else if (ctx->state == STATE_PATH && ctx->buf[ctx->pos] == '/') {
+		// new path token
+		ctx->buf[ctx->pos - ctx->shift] = '\0';
+		ls_push (ctx->path, (void *)ctx->token_begin);
+		ctx->token_begin = ctx->pos + 1;
+		ctx->shift = 0;
+	} else if (ctx->state == STATE_KEY && ctx->buf[ctx->pos] == '=') {
+		// switch from key into value mode
+		ctx->buf[ctx->pos - ctx->shift] = '\0';
+		ctx->token_begin = ctx->pos + 1;
+		ctx->shift = 0;
+		ctx->state = STATE_VALUE;
+	} else if (ctx->shift) {
+		// just some char, shift it back if necessary
+		ctx->buf[ctx->pos - ctx->shift] = ctx->buf[ctx->pos];
+	}
+	ctx->pos++;
+}
+
+#define INITIAL_BUFSZ 32
+
+SDB_API bool sdb_text_load_fd(Sdb *s, int fd) {
+	LoadCtx ctx = {
+		.fd = fd,
+		.eof = false,
+		.bufsz = INITIAL_BUFSZ,
+		.buf = malloc (INITIAL_BUFSZ),
+		.root_db = s,
+		.cur_db = s,
+		.pos = 0,
+		.token_begin = 0,
+		.shift = 0,
+		.path = ls_new (),
+		.state = STATE_NEWLINE,
+		.unescape = false
+	};
+	if (!ctx.buf || !ctx.path) {
```

Comment:
if (ctx. buf && ctx.path) { ...} so you avoid having dupplicated free/ls_free. btw thise handmade free and struct initialization can be done in a separate functions, so it makes it cleaner, mainly because ctx just needs the fd to be passed, all the other stuff is 0 or allocated